File: riveredge-frontend/src/apps/kuaizhizao/services/quotation.ts
/**
 * 报价单管理API服务
 *
 * 提供报价单相关的API调用接口。
 *
 * @author RiverEdge Team
 * @date 2026-02-19
 */

import { apiRequest } from '../../../services/api';
import type { DocumentPrintApiResult } from '../../../utils/printResponseHelpers';

export interface ActionCapability {
  allowed: boolean;
  reason?: string | null;
}

export interface QuotationCapabilities {
  update: ActionCapability;
  delete: ActionCapability;
  submit: ActionCapability;
  withdraw_submit: ActionCapability;
  approve: ActionCapability;
  revoke_approval: ActionCapability;
  confirm_customer: ActionCapability;
  cancel_customer_confirm: ActionCapability;
  convert_to_order: ActionCapability;
  convert_to_contract: ActionCapability;
  revoke_push: ActionCapability;
  reopen: ActionCapability;
  create_revision: ActionCapability;
  print_formal: ActionCapability;
}

export interface QuotationItem {
  id?: number;
  uuid?: string;
  tenant_id?: number;
  quotation_id?: number;
  material_id?: number;
  material_code?: string;
  material_name?: string;
  material_spec?: string;
  material_unit?: string;
  quote_quantity?: number;
  unit_price?: number;
  /** 税率（%），含税模式下录入 */
  tax_rate?: number;
  total_amount?: number;
  delivery_date?: string;
  notes?: string;
  created_at?: string;
  updated_at?: string;
}

export interface Quotation {
  id?: number;
  uuid?: string;
  tenant_id?: number;
  quotation_code?: string;
  quotation_date?: string;
  valid_until?: string;
  delivery_date?: string;
  customer_id?: number;
  customer_name?: string;
  customer_contact?: string;
  customer_phone?: string;
  total_quantity?: number;
  total_amount?: number;
  /** 整单优惠金额（从价税合计扣减，不改明细单价） */
  discount_amount?: number;
  /** 与销售订单一致：tax_inclusive 含税单价 / tax_exclusive 不含税单价 */
  price_type?: 'tax_inclusive' | 'tax_exclusive' | string;
  status?: string;
  reviewer_id?: number;
  reviewer_name?: string;
  review_time?: string;
  review_status?: string;
  review_remarks?: string;
  salesman_id?: number;
  salesman_name?: string;
  shipping_address?: string;
  shipping_method?: string;
  payment_terms?: string;
  /** 币种（数据字典 CURRENCY，默认 CNY 人民币） */
  currency_code?: string;
  sales_order_id?: number;
  sales_order_code?: string;
  contract_id?: number;
  contract_code?: string;
  /** 报价系列编码（首版通常与 quotation_code 相同） */
  quotation_series_code?: string;
  root_quotation_id?: number;
  version_no?: number;
  previous_quotation_id?: number;
  is_latest_in_series?: boolean;
  superseded_by_id?: number;
  formal_document_generated_at?: string;
  /** 已转单但下游销售订单已不存在（如已删除） */
  conversion_downstream_missing?: boolean;
  /** 仍有关联 contract_id，但下游销售合同已不存在（如已删除） */
  contract_downstream_missing?: boolean;
  notes?: string;
  is_active?: boolean;
  created_by?: number;
  updated_by?: number;
  created_at?: string;
  updated_at?: string;
  items?: QuotationItem[];
  lifecycle?: Record<string, unknown>;
  capabilities?: QuotationCapabilities;
}

export interface QuotationListParams {
  skip?: number;
  limit?: number;
  status?: string;
  salesman_id?: number;
  start_date?: string;
  end_date?: string;
  /** 工具栏模糊搜索：编号、客户 */
  keyword?: string;
  quotation_code?: string;
  customer_name?: string;
  quotation_series_code?: string;
  /** 数据范围：all 全部 / mine 我的 / department 我的部门 */
  list_scope?: 'all' | 'mine' | 'department';
  /** 仅可上拉建销售订单：未关联销售订单且非已转订单 */
  pullable_only?: boolean;
  /** 上拉目标单据类型：销售订单 / 销售合同 */
  pull_target?: 'sales_order' | 'sales_contract';
}

export interface QuotationListResponse {
  data: Quotation[];
  total: number;
  success: boolean;
}

export async function listQuotations(params: QuotationListParams = {}): Promise<QuotationListResponse> {
  return apiRequest<QuotationListResponse>('/apps/kuaizhizao/quotations', {
    method: 'GET',
    params,
  });
}

export async function getQuotation(id: number, includeItems: boolean = true): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}`, {
    method: 'GET',
    params: { include_items: includeItems },
  });
}

export async function createQuotation(
  data: Partial<Quotation>,
  options?: { autoSubmit?: boolean },
): Promise<Quotation> {
  return apiRequest<Quotation>('/apps/kuaizhizao/quotations', {
    method: 'POST',
    data,
    params:
      options?.autoSubmit === false
        ? { auto_submit: false }
        : options?.autoSubmit === true
          ? { auto_submit: true }
          : undefined,
  });
}

export async function updateQuotation(id: number, data: Partial<Quotation>): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}`, {
    method: 'PUT',
    data,
  });
}

/** 提交报价单（草稿 → 已发送；蓝图无需审核时自动通过） */
export async function submitQuotation(id: number): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}/submit`, {
    method: 'POST',
  });
}

/** 撤回报价单（已发送 + 待审核 → 草稿） */
export async function withdrawQuotation(id: number): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}/withdraw`, {
    method: 'POST',
  });
}

/** 审核通过（已发送 + 待审核 → 保持已发送，审核通过） */
export async function approveQuotation(id: number, data?: { review_remarks?: string }): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}/approve`, {
    method: 'POST',
    data: data ?? {},
  });
}

/** 驳回（已发送 + 待审核 → 已拒绝） */
export async function rejectQuotation(id: number, data?: { review_remarks?: string }): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}/reject`, {
    method: 'POST',
    data: data ?? {},
  });
}

/** 撤回审核（已发送 + 已通过 → 待审核） */
export async function revokeReviewQuotation(id: number): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}/revoke-review`, {
    method: 'POST',
  });
}

/** 客户确认 / 标记已接受（已发送 + 已通过 → 已接受） */
export async function confirmCustomerQuotation(id: number): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}/confirm-customer`, {
    method: 'POST',
  });
}

/** 客户取消确认（已接受 → 已发送，可撤回审核或删除） */
export async function cancelCustomerConfirmQuotation(id: number): Promise<Quotation> {
  return apiRequest<Quotation>(
    `/apps/kuaizhizao/quotations/${id}/cancel-customer-confirm`,
    { method: 'POST' },
  );
}

/** 驳回后重新编辑（已拒绝 → 草稿） */
export async function reopenQuotation(id: number): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}/reopen`, {
    method: 'POST',
  });
}

/** 撤回下推（已转订单且下游销售订单已不存在 → 已接受） */
export async function revokePushQuotation(id: number): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${id}/revoke-push`, {
    method: 'POST',
  });
}

/** 新建修订版：从系列当前最新版复制为草稿修订 */
export async function createQuotationRevision(
  quotationId: number,
  data?: Record<string, unknown>
): Promise<Quotation> {
  return apiRequest<Quotation>(`/apps/kuaizhizao/quotations/${quotationId}/revision`, {
    method: 'POST',
    data: data ?? {},
  });
}

export async function deleteQuotation(id: number): Promise<void> {
  return apiRequest<void>(`/apps/kuaizhizao/quotations/${id}`, {
    method: 'DELETE',
  });
}

export interface ConvertToOrderResponse {
  sales_order: { id?: number; order_code?: string; [key: string]: any };
  quotation: Quotation;
}

export async function convertQuotationToOrder(quotationId: number): Promise<ConvertToOrderResponse> {
  return apiRequest<ConvertToOrderResponse>(`/apps/kuaizhizao/quotations/${quotationId}/convert-to-order`, {
    method: 'POST',
  });
}

export async function printQuotation(
  quotationId: number,
  options?: {
    templateUuid?: string;
    outputFormat?: 'html' | 'pdf';
    responseFormat?: 'json' | 'html';
  }
): Promise<DocumentPrintApiResult> {
  return apiRequest<DocumentPrintApiResult>(`/apps/kuaizhizao/quotations/${quotationId}/print`, {
    method: 'GET',
    params: {
      template_uuid: options?.templateUuid,
      output_format: options?.outputFormat ?? 'pdf',
      response_format: options?.responseFormat ?? 'json',
    },
  });
}

export async function getQuotationPrintVariables(quotationId: number): Promise<Record<string, any>> {
  return apiRequest<Record<string, any>>(`/apps/kuaizhizao/quotations/${quotationId}/print-variables`, {
    method: 'GET',
  });
}

export async function recordQuotationPrint(quotationId: number): Promise<{ success: boolean; timestamp?: string }> {
  return apiRequest<{ success: boolean; timestamp?: string }>(`/apps/kuaizhizao/quotations/${quotationId}/record-print`, {
    method: 'POST',
  });
}

File: riveredge-backend/src/apps/kuaizhizao/schemas/quotation.py
"""
报价单管理模块数据验证schema

提供报价单相关的数据验证和序列化。

Author: RiverEdge Team
Date: 2026-02-19
"""

from __future__ import annotations

from datetime import datetime, date
from typing import Optional, List, Dict, Any
from decimal import Decimal
from pydantic import Field, model_validator
from core.schemas.base import BaseSchema


# === 报价单明细 ===

class QuotationItemBase(BaseSchema):
    """报价单明细基础schema"""
    material_id: int = Field(..., description="物料ID")
    material_code: str = Field(..., max_length=100, description="物料编码")
    material_name: str = Field(..., max_length=200, description="物料名称")
    material_spec: Optional[str] = Field(None, max_length=200, description="物料规格")
    material_unit: str = Field(..., max_length=20, description="物料单位")
    quote_quantity: Decimal = Field(..., gt=0, description="报价数量")
    unit_price: Decimal = Field(..., ge=0, description="单价")
    tax_rate: Decimal = Field(Decimal("0"), ge=0, le=100, description="税率（%）")
    total_amount: Optional[Decimal] = Field(None, ge=0, description="金额")
    variant_attributes: Optional[Dict[str, Any]] = Field(None, description="属性组合（临时组合）")
    delivery_date: Optional[date] = Field(None, description="预计交货日期")
    notes: Optional[str] = Field(None, description="备注")


class QuotationItemCreate(QuotationItemBase):
    """创建报价单明细schema"""
    pass


class QuotationItemUpdate(BaseSchema):
    """更新报价单明细schema"""
    material_id: Optional[int] = None
    material_code: Optional[str] = Field(None, max_length=100)
    material_name: Optional[str] = Field(None, max_length=200)
    material_spec: Optional[str] = Field(None, max_length=200)
    material_unit: Optional[str] = Field(None, max_length=20)
    quote_quantity: Optional[Decimal] = Field(None, gt=0)
    unit_price: Optional[Decimal] = Field(None, ge=0)
    tax_rate: Optional[Decimal] = Field(None, ge=0, le=100)
    total_amount: Optional[Decimal] = Field(None, ge=0)
    variant_attributes: Optional[Dict[str, Any]] = None
    delivery_date: Optional[date] = None
    notes: Optional[str] = None


class QuotationItemResponse(QuotationItemBase):
    """报价单明细响应schema"""
    id: int
    uuid: str
    tenant_id: int
    quotation_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# === 报价单 ===

class QuotationBase(BaseSchema):
    """报价单基础schema"""
    quotation_code: Optional[str] = Field(None, max_length=120, description="报价单编码（自动生成；修订版含 -Vn 后缀）")
    quotation_date: date = Field(..., description="报价日期")
    valid_until: Optional[date] = Field(None, description="有效期至")
    delivery_date: Optional[date] = Field(None, description="预计交货日期")

    # 客户信息（必填）
    customer_id: int = Field(..., description="客户ID")
    customer_name: str = Field(..., max_length=200, description="客户名称")
    customer_contact: Optional[str] = Field(None, max_length=100, description="客户联系人")
    customer_phone: Optional[str] = Field(None, max_length=20, description="客户电话")

    # 金额信息
    total_quantity: Decimal = Field(Decimal("0"), ge=0, description="总数量")
    total_amount: Decimal = Field(Decimal("0"), ge=0, description="总金额")
    price_type: Optional[str] = Field(
        "tax_exclusive",
        max_length=20,
        description="价格类型：含税(tax_inclusive)/不含税(tax_exclusive)",
    )

    # 状态：草稿/已发送/已接受/已拒绝/已转订单
    status: str = Field("草稿", max_length=20, description="报价状态")

    # 审核信息
    reviewer_id: Optional[int] = Field(None, description="审核人ID")
    reviewer_name: Optional[str] = Field(None, max_length=100, description="审核人姓名")
    review_time: Optional[datetime] = Field(None, description="审核时间")
    review_status: Optional[str] = Field("待审核", max_length=20, description="审核状态")
    review_remarks: Optional[str] = Field(None, description="审核备注")

    # 销售信息
    salesman_id: Optional[int] = Field(None, description="销售员ID")
    salesman_name: Optional[str] = Field(None, max_length=100, description="销售员姓名")

    # 物流信息
    shipping_address: Optional[str] = Field(None, description="收货地址")
    shipping_method: Optional[str] = Field(None, max_length=50, description="发货方式")
    payment_terms: Optional[str] = Field(None, max_length=100, description="付款条件")
    currency_code: Optional[str] = Field("CNY", max_length=20, description="币种代码（默认 CNY 人民币）")

    notes: Optional[str] = Field(None, description="备注")


class QuotationCreate(QuotationBase):
    """创建报价单schema"""
    items: List[QuotationItemCreate] = Field(default_factory=list, description="报价明细")

    @model_validator(mode='after')
    def validate_items(self):
        """验证报价明细"""
        if not self.items or len(self.items) == 0:
            raise ValueError("报价单必须至少包含一条明细")
        return self


class QuotationUpdate(BaseSchema):
    """更新报价单schema"""
    quotation_date: Optional[date] = None
    valid_until: Optional[date] = None
    delivery_date: Optional[date] = None
    customer_id: Optional[int] = None
    customer_name: Optional[str] = Field(None, max_length=200)
    customer_contact: Optional[str] = Field(None, max_length=100)
    customer_phone: Optional[str] = Field(None, max_length=20)
    total_quantity: Optional[Decimal] = Field(None, ge=0)
    total_amount: Optional[Decimal] = Field(None, ge=0)
    price_type: Optional[str] = Field(None, max_length=20)
    status: Optional[str] = Field(None, max_length=20)
    salesman_id: Optional[int] = None
    salesman_name: Optional[str] = Field(None, max_length=100)
    shipping_address: Optional[str] = None
    shipping_method: Optional[str] = Field(None, max_length=50)
    payment_terms: Optional[str] = Field(None, max_length=100)
    currency_code: Optional[str] = Field(None, max_length=20)
    notes: Optional[str] = None
    items: Optional[List[QuotationItemCreate]] = None


class QuotationResponse(QuotationBase):
    """报价单响应schema"""
    id: int
    uuid: str
    tenant_id: int
    quotation_series_code: Optional[str] = Field(None, max_length=120, description="报价系列编码")
    root_quotation_id: Optional[int] = Field(None, description="系列根报价单 ID")
    version_no: int = Field(1, description="系列内版本号")
    previous_quotation_id: Optional[int] = Field(None, description="上一版本报价单 ID")
    is_latest_in_series: bool = Field(True, description="是否为系列最新版本")
    superseded_by_id: Optional[int] = Field(None, description="被替代为的新版本 ID")
    formal_document_generated_at: Optional[datetime] = Field(
        None, description="首次生成正式报价 PDF 的时间"
    )
    sales_order_id: Optional[int] = Field(None, description="关联销售订单ID")
    sales_order_code: Optional[str] = Field(None, max_length=50, description="关联销售订单编码")
    is_active: bool = Field(True, description="是否有效")
    created_by: Optional[int] = None
    updated_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    items: Optional[List[QuotationItemResponse]] = Field(None, description="报价明细")
    lifecycle: Optional[dict] = Field(None, description="生命周期（后端计算，供 UniLifecycleStepper 展示）")
    conversion_downstream_missing: Optional[bool] = Field(
        None,
        description="已转单标记下，下游销售订单已不存在（如已删除）时为 True，可重新下推或删除本报价单",
    )

    class Config:
        from_attributes = True


class QuotationListResponse(BaseSchema):
    """报价单列表响应schema"""
    data: List[QuotationResponse]
    total: int
    success: bool = True


class QuotationReviewAction(BaseSchema):
    """报价单审核类操作可选备注"""
    review_remarks: Optional[str] = Field(None, max_length=500, description="审核意见/备注")


class QuotationRevisionBody(BaseSchema):
    """新建修订版：可选覆盖头字段与明细；未传则复制系列当前最新版。"""

    quotation_date: Optional[date] = None
    valid_until: Optional[date] = None
    delivery_date: Optional[date] = None
    customer_id: Optional[int] = None
    customer_name: Optional[str] = Field(None, max_length=200)
    customer_contact: Optional[str] = Field(None, max_length=100)
    customer_phone: Optional[str] = Field(None, max_length=20)
    total_quantity: Optional[Decimal] = Field(None, ge=0)
    total_amount: Optional[Decimal] = Field(None, ge=0)
    price_type: Optional[str] = Field(None, max_length=20)
    salesman_id: Optional[int] = None
    salesman_name: Optional[str] = Field(None, max_length=100)
    shipping_address: Optional[str] = None
    shipping_method: Optional[str] = Field(None, max_length=50)
    payment_terms: Optional[str] = Field(None, max_length=100)
    currency_code: Optional[str] = Field(None, max_length=20)
    notes: Optional[str] = None
    items: Optional[List[QuotationItemCreate]] = None
